### PR TITLE
Beta indicator

### PR DESCRIPTION
**Pull request by Marinovsky**
In technical analysis Beta indicator is used to measure volatility or risk of a stock (ETF) relative to the overall risk (volatility) of the stock market (market indexes). The Beta indicators compares stock's price movement to the movements of the indexes over the same period of time.

It is common practice to use the S&P 500 index as a benchmark of the overall stock market when it comes to Beta calculations.

The concept of the Beta analysis is quite simple.

 - If stock's movements are more volatile than the overall market, the Beta values will be garter than 1;
 - When stock is less volatile, the Beta values are less than 1;
 - When stock's movement are inversed to the market movements, the Beta values will be negative with the same logic that more volatile stocks will have absolute Beta values above 1 and less volatile stocks will have absolute Beta values below 1.
 
Beta indicator example with IBM and SPY stocks with a period of 15 days from January 1th 2019 to June 1th 2020
image
https://www.quantconnect.com/terminal/processCache?request=embedded_backtest_e2094b44a575c60f200a5eb3f1f84c29.html
Beta indicator from TradingView with the same parameters
image

#### Description

- Create `BetaIndicator.cs`
- Create `B` method in `QCAlgorithm.Indicators`
- Add unit tests in `BetaIndicatorTests.cs`
- Add unit test file `bi_datatest.csv`
- Add regression test
#### Related Issue

#83
#### Motivation and Context

With this change the user will be able to get the beta of a stock in relation with a market index within certain rolling period of time
#### Requires Documentation Change

This change require an update to the documentation describing the new indicator and how to use it
#### How Has This Been Tested?

- It was made a test file with the expected beta value of AMZN stock in relation with market index S&P 500. The beta was calculated using Microsoft Excel and the historical data, from AMZN and S&P 500, was extracted from Yahoo finance. Finally, from October 4th to November 5th of 2021 it was asserted the expected Beta values with the Beta values computed by the indicator
- A regression test was conducted where a new beta indicator was created and stated that its values were between a lower and upper limit. In the regression test, when the beta between IBM and SPY was close to 1 and SPY values were increasing, the shares purchased from IBM were sold
#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm/QCAlgorithm.Indicators.cs [hunk 1]
```diff
@@ -352,6 +352,31 @@ public BollingerBands BB(Symbol symbol, int period, decimal k, MovingAverageType
...
+            RegisterIndicator(target, b, resolution);
+            RegisterIndicator(reference, b, resolution);
+
+            if (EnableAutomaticIndicatorWarmUp)
+            {
+                WarmUpIndicator(target, b, resolution);
+                WarmUpIndicator(reference, b, resolution);
+            }
+
+            return b;
```
Comment (Martin-Molinero) [inline Algorithm/QCAlgorithm.Indicators.cs:377]: nit - `b` -> `betaIndicator`

File: Indicators/BetaIndicator.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        /// using the target and reference returns
+        /// </summary>
+        private void ComputeBeta()
+        {
+            var targetReturnsList = _targetReturns.ToList();
+            var referenceReturnsList = _referenceReturns.ToList();
+
+            // Avoid division with NaN or by zero
+            var variance = !referenceReturnsList.Variance().IsNaNOrZero() ? referenceReturnsList.Variance() : 1;
+            var covariance = !targetReturnsList.Covariance(referenceReturnsList).IsNaNOrZero() ? targetReturnsList.Covariance(referenceReturnsList) : 0;
```
Comment (Martin-Molinero) [inline Indicators/BetaIndicator.cs:153]: - Your calculating the variance and covariance twice here, should calculate it once and store its value
- No need for the `ToList()` calls above I think?

File: Indicators/BetaIndicator.cs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+        /// <summary>
+        /// Computes the returns with the new given data point and the last given data point
+        /// </summary>
+        /// <param name="rollingWindow">The collection of data points from which we want
+        /// to compute the return</param>
+        /// <returns>The returns with the new given data point</returns>
+        private static double GetNewReturn(RollingWindow<decimal> rollingWindow)
+        {
+            return (double) ((rollingWindow[1] / rollingWindow[0]) - 1);
```
Comment (Martin-Molinero) [inline Indicators/BetaIndicator.cs:139]: Hm you have a rolling window of size `WarmUpPeriod` but only really using the last 2 values in it?

File: Indicators/BetaIndicator.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        /// <returns>The beta value of the target used in relation with the reference</returns>
+        protected override decimal ComputeNextValue(TradeBar input)
+        {
+            var inputSymbol = input.Symbol;
+            _symbolDataPoints[inputSymbol].Add(input.Close);
+
+            if (_symbolDataPoints[_targetSymbol].Samples == _symbolDataPoints[_referenceSymbol].Samples && _symbolDataPoints[_targetSymbol].Samples > 1)
+            {
+                _targetReturns.Add(GetNewReturn(_symbolDataPoints[_targetSymbol]));
+                _referenceReturns.Add(GetNewReturn(_symbolDataPoints[_referenceSymbol]));
```
Comment (Martin-Molinero) [inline Indicators/BetaIndicator.cs:124]: Suggest storing `_symbolDataPoints[_referenceSymbol]` & `_symbolDataPoints[_targetSymbol]` in a local variable 👍

**Review by Martin-Molinero — APPROVED** [review]
Cool! 🍬 minor comments

File: Indicators/BetaIndicator.cs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+        /// <summary>
+        /// Creates a new BetaIndicator with the specified name, period, target and 
+        /// reference values
+        /// </summary>
+        /// <param name="name">The name of this indicator</param>
+        /// <param name="period">The period of this indicator</param>
+        /// <param name="targetSymbol">The target symbol of this indicator</param>
+        /// <param name="referenceSymbol">The reference symbol of this indicator</param>
+        public BetaIndicator(string name, int period, Symbol targetSymbol, Symbol referenceSymbol)
```
Comment (jaredbroad) [inline Indicators/BetaIndicator.cs:85]: Mind renaming this to be inline with the other indicators? e.g. `ExponentialMovingAverage.cs` / so `Beta.cs`

**Review by jaredbroad — COMMENTED** [review]
So cool to finally see BETA indicator! Minor renaming change

File: Indicators/BetaIndicator.cs (hunk 5 above)
Comment (Marinovsky) [reply Indicators/BetaIndicator.cs:85]: But of course. No problem!

**Pull request merged**